networkx/algorithms/isomorphism/vf2pp.py

```diff
@@ -0,0 +1,302 @@
+            if stack:
+                # Pop the previously added u-v pair, and look for a different candidate _v for u
+                popped_node1, _ = stack[-1]
+                popped_node2 = mapping[popped_node1]
+                mapping.pop(popped_node1)
+                reverse_mapping.pop(popped_node2)
+                _restore_Tinout(popped_node1, popped_node2, graph_params, state_params)
+            continue
+
+        if _feasibility(current_node, candidate, graph_params, state_params):
+            # Terminate if mapping is extended to its full
+            if len(mapping) == G2.number_of_nodes() - 1:
+                cp_mapping = mapping.copy()
+                cp_mapping[current_node] = candidate
+                yield cp_mapping
+                continue
+
+            # Feasibility rules pass, so extend the mapping and update the parameters
+            mapping.update({current_node: candidate})
+            reverse_mapping.update({candidate: current_node})
+            _update_Tinout(current_node, candidate, graph_params, state_params)
+            # Append the next node and its candidates to the stack
+            candidates = iter(
+                _find_candidates(node_order[matching_node], graph_params, state_params)
+            )
+            stack.append((node_order[matching_node], candidates))
+            matching_node += 1
+
+
+def _precheck(G1, G2, G1_labels, G2_labels, node_labels=None, default_label=-1):
+    """Checks if all the pre-requisites are satisfied before calling the isomorphism solver.
+
+    Notes
+    -----
+    Before trying to create the mapping between the nodes of the two graphs, we must check if:
+    1. The two graphs have equal number of nodes
+    2. The degree sequences in the two graphs are identical
+    3. The two graphs have the same label distribution. For example, if G1 has orange nodes but G2 doesn't, there's no
+    point in proceeding to create a mapping.
+
+    Parameters
+    ----------
+    G1,G2: NetworkX Graph or MultiGraph instances.
+        The two graphs to check for isomorphism or monomorphism.
+
+    G1_labels,G2_labels: dict
+        The label of every node in G1 and G2 respectively.
+
+        node_labels: Label name
+        The label name of all nodes
+
+    default_label: Label name
+        Let the user pick a default label value
+    """
+    if G1.order() != G2.order():
+        return False
+    if sorted(d for n, d in G1.degree()) != sorted(d for n, d in G2.degree()):
+        return False
+
+    G1_labels.update(G1.nodes(data=node_labels, default=default_label))
+    G2_labels.update(G2.nodes(data=node_labels, default=default_label))
+
+    G1_nodes_per_label = {
+        label: len(nodes) for label, nodes in nx.utils.groups(G1_labels).items()
+    }
+
+    if any(
+        label not in G1_nodes_per_label or G1_nodes_per_label[label] != len(nodes)
+        for label, nodes in nx.utils.groups(G2_labels).items()
+    ):
+        return False
```

Comment:
This code in `_precheck` calculates the label groups. Then we recalculate them in `_initialize_VF2pp`.
I've been debating tradeoffs between doing the prechecks before setting up the data structures or vice versa or some mix of the two. Realizing that this precheck relies on the label groups, and thus on the label dicts makes me feel more strongly that we should set up the data structures before the prechecks.

It is true that some of the prechecks don't require anything to do with labels. so it will (in cases where the graphs are obviously not isomorphic) be slower to return False if we manipulate the labels before checking. But it still will be quite fast.  If that is a concern, another formulation that might work well would be:
```python
    if not nx.faster_isomorphism_check(G1, G2):
        return False
    graph_params, state_params = _initialize_VF2pp(...)
    _precheck_of_label_properties(...)
```
This would put use the already exiting checks inside `faster_isomorphism_check` which we should check to make sure it is what we want. The important thing for me is to organize the code so we don't compute something twice and so it is easy to find where the data structures are created.